Selector pattern parameterized

EdwardALuke opened the pull request:
This update adds a parametric version of the selector pattern.  This pattern uses parametric rules to automate the process of generating constraints and validating inputs used to select different methods using constraints.  As an example of how this is used the limiter infrastructure is updated to use the new selector pattern.  This only changes the behavior of the limiter in that it will error out if the limiter is set to an invalid selection instead of defaulting to venkatakrishnan, which is probably a more desirable outcome.

One advantage of the new patterns is that it is now possible to create third party add on modules that can add new limiters without needing to change the FVMMod.  

The way that this pattern is you need to create a default rule to define the default setting of the selector (which must be of type param<string>)  Then for each selected algorithm you add a rule such as

 $rule pointwise(OUTPUT<-selector(limiter,venkatakrishnan,V)) {}

Note, this rule is not actually executed, but it causes rules to be instantiated to manage the creation of code that will check and create a constraint for the selector and also create validators that will check that one and only one of the constraints are set to ~EMPTY.  The first argument of the selector parametric variable is the input variable for the selector, the second is the selection name, and the third is an abbreviation of the selected name.  In this case when limiter is set to either "venkatakrishnan" or "V" then the constraint limiter_select_venkatakrishnan will be set to UNIVERSE (~EMPTY).  All other selector generated constraints will be empty.

EdwardALuke committed:
Added new selector infrastructure to standardized selection idiom

EdwardALuke committed:
Update to change limiters to use new selector infrastructure

TimothyEDawson commented:
@EdwardALuke I really like this, both in how it simplifies a very common pattern and how it moves the logic for adding new options closer to the code which implements those options! I expect it will be very nice for the time integration options in CHEM, and especially for the custom time integrators I've implemented using modules in the past.

One question: In this current implementation, is the abbreviation an optional or a required input?

EdwardALuke commented:
@TimothyEDawson As it is implemented the abbreviation is required, but I think we could support specifying no abbreviation as an option just using a two parameter parametric variable.   I will work on making those changes.

EdwardALuke committed:
Update to add option to create selector without abbreviation

EdwardALuke commented:
@TimothyEDawson  Ok, I have changed the code so that you can specify a selector with or without an abbreviation.  I have also modified the none and zero limiters so that they no longer specify an abbreviation using the new API.

EdwardALuke committed:
Update centroid selector to use new API

EdwardALuke committed:
Updated faceGradStencil to use new selector API

EdwardALuke committed:
Updated cell gradient selector to use new selector API.

EdwardALuke commented:
I am going to merge this into dev so I can start getting CHEM to conform to the updated API.
